tensorflow/core/kernels/data/batch_dataset_op_test.cc

```diff
@@ -51,677 +52,973 @@ class BatchDatasetOpTest : public DatasetOpsTestBase {
-          /*expected_output_shapes*/ {PartialTensorShape({3})},
-          /*expected_cardinality*/ 3,
-          /*breakpoints*/ {0, 1, 5}};
-}
-
-class ParameterizedBatchDatasetOpTest
-    : public BatchDatasetOpTest,
-      public ::testing::WithParamInterface<TestCase> {};
+BatchDatasetParams InvalidBatchSizeBatchDataset() {
+  return {/*start=*/0,
+          /*stop=*/10,
+          /*step=*/1,
+          /*batch_size=*/-1,
+          /*drop_remainder=*/false,
+          /*parallel_copy=*/false,
+          /*output_dtypes=*/{DT_INT64},
+          /*output_shapes=*/{PartialTensorShape({3})},
+          /*node_name=*/"batch_dataset"};
+}
+
+class ParameterizedGetNextTest : public BatchDatasetOpTest,
+                                 public ::testing::WithParamInterface<
+                                     GetNextTestCase<BatchDatasetParams>> {};
+
+GetNextTestCase<BatchDatasetParams> GetNextTestCase1() {
+  return {/*dataset_params=*/BatchDataset1(),
+          /*expected_outputs=*/
+          CreateTensors<int64>(TensorShape({4}),
+                               {{0, 1, 2, 3}, {4, 5, 6, 7}, {8, 9, 10, 11}})};
+}
+
+GetNextTestCase<BatchDatasetParams> GetNextTestCase2() {
+  return {/*dataset_params=*/BatchDataset2(),
+          /*expected_outputs=*/
+          CreateTensors<int64>(TensorShape({4}),
+                               {{0, 1, 2, 3}, {4, 5, 6, 7}, {8, 9, 10, 11}})};
+}
+
+GetNextTestCase<BatchDatasetParams> GetNextTestCase3() {
+  return {/*dataset_params=*/BatchDataset3(),
+          /*expected_outputs=*/
+          {CreateTensor<int64>(TensorShape({3}), {0, 1, 2}),
+           CreateTensor<int64>(TensorShape({3}), {3, 4, 5}),
+           CreateTensor<int64>(TensorShape({3}), {6, 7, 8}),
+           CreateTensor<int64>(TensorShape({1}), {9})}};
+}
+
+GetNextTestCase<BatchDatasetParams> GetNextTestCase4() {
+  return {/*dataset_params=*/BatchDataset4(),
+          /*expected_outputs=*/
+          CreateTensors<int64>(TensorShape({3}),
+                               {{0, 1, 2}, {3, 4, 5}, {6, 7, 8}})};
+}
+
+GetNextTestCase<BatchDatasetParams> GetNextTestCase5() {
+  return {/*dataset_params=*/BatchDataset5(),
+          /*expected_outputs=*/{}};
+}
 
-TEST_P(ParameterizedBatchDatasetOpTest, GetNext) {
+GetNextTestCase<BatchDatasetParams> GetNextTestCase6() {
+  return {/*dataset_params=*/BatchDataset6(),
+          /*expected_outputs=*/
+          CreateTensors<int64>(TensorShape({10}),
+                               {{0, 1, 2, 3, 4, 5, 6, 7, 8, 9}})};
+}
+
+GetNextTestCase<BatchDatasetParams> GetNextTestCase7() {
+  return {/*dataset_params=*/BatchDataset7(),
+          /*expected_outputs=*/{}};
+}
+
+TEST_P(ParameterizedGetNextTest, GetNext) {
   int thread_num = 2, cpu_num = 2;
-  TestCase test_case = GetParam();
   TF_ASSERT_OK(InitThreadPool(thread_num));
   TF_ASSERT_OK(InitFunctionLibraryRuntime({}, cpu_num));
 
+  auto test_case = GetParam();
+
   std::unique_ptr<OpKernel> batch_dataset_kernel;
```

Comment:
After we make the unit test cases more modular, the unit test becomes more concise and also modular. As an example, the following code block used to create `dataset_kernel`, `dataset_context`, `dataset`, `iterator_context`, and `iterator` is repeated in each unit function test. Do you think we need to propose a solution to avoid the repeat code?

```C++
std::unique_ptr<OpKernel> batch_dataset_kernel;
  TF_ASSERT_OK(CreateBatchDatasetOpKernel(
      test_case.dataset_params.parallel_copy,
      test_case.dataset_params.output_dtypes,
      test_case.dataset_params.output_shapes, &batch_dataset_kernel));

  TF_ASSERT_OK(MakeRangeDataset(
      test_case.dataset_params.range_dataset_params.start,
      test_case.dataset_params.range_dataset_params.stop,
      test_case.dataset_params.range_dataset_params.step,
      test_case.dataset_params.range_dataset_params.output_dtypes,
      test_case.dataset_params.range_dataset_params.output_shapes,
      &test_case.dataset_params.input_dataset));

  gtl::InlinedVector<TensorValue, 4> inputs;
  TF_ASSERT_OK(test_case.dataset_params.MakeInputs(&inputs));

  std::unique_ptr<OpKernelContext> batch_dataset_context;
  TF_ASSERT_OK(CreateBatchDatasetContext(batch_dataset_kernel.get(), &inputs,
                                         &batch_dataset_context));
  DatasetBase* batch_dataset;
  TF_ASSERT_OK(CreateDataset(batch_dataset_kernel.get(),
                             batch_dataset_context.get(), &batch_dataset));
  core::ScopedUnref scoped_unref(batch_dataset);

  std::unique_ptr<IteratorContext> iterator_ctx;
  TF_ASSERT_OK(
      CreateIteratorContext(batch_dataset_context.get(), &iterator_ctx));
  std::unique_ptr<IteratorBase> iterator;
  TF_ASSERT_OK(batch_dataset->MakeIterator(iterator_ctx.get(), kIteratorPrefix,
                                           &iterator));
```

Reply:
Will submit a commit for discussion soon.

Reply:
sounds good